File: agent_skills/biomarker_discovery.py
"""Agent skill: End-to-end biomarker discovery workflow.

Orchestrates the full pipeline from data loading through feature explanation.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable


class BiomarkerDiscoverySkill:
    """Orchestrate biomarker discovery across the full MCP tool chain."""

    def __init__(self, tool_caller: Callable | None = None) -> None:
        self.tool_caller = tool_caller

    async def _call_tool(self, tool_name: str, **kwargs: Any) -> dict:
        """Call an MCP tool or raise if no caller configured."""
        if self.tool_caller is None:
            raise RuntimeError("No tool_caller configured for BiomarkerDiscoverySkill")
        result = self.tool_caller(tool_name, **kwargs)
        if hasattr(result, "__await__"):
            result = await result
        return result

    async def run(
        self,
        target: str = "msi",
        modalities: list[str] | None = None,
        config: dict | None = None,
    ) -> dict:
        """Orchestrate: load -> impute -> availability -> select -> classify -> match -> explain -> report."""
        modalities = modalities or ["proteomics", "rnaseq"]
        config = config or {}
        report: dict[str, Any] = {"target": target, "modalities": modalities, "steps": {}}

        # Step 1: Load dataset
        load_result = await self._call_tool(
            "load_dataset",
            dataset=config.get("dataset", "train"),
            modalities=["clinical"] + modalities,
            data_dir=config.get("data_dir"),
        )
        report["steps"]["load_dataset"] = load_result

        # Step 2: Impute missing values (per modality)
        impute_results: dict[str, Any] = {}
        for modality in modalities:
            impute_result = await self._call_tool(
                "impute_missing_values",
                dataset=config.get("dataset", "train"),
                modality=modality,
                strategy=config.get("impute_strategy", "nmf"),
                classify_missingness=True,
            )
            impute_results[modality] = impute_result
        report["steps"]["impute_missing"] = impute_results

        # Step 3: Availability check
        avail_result = await self._call_tool(
            "check_availability",
            threshold=config.get("availability_threshold", 0.9),
            dataset=config.get("dataset", "train"),
            use_imputed=True,
        )
        report["steps"]["check_availability"] = avail_result

        # Step 4: Select biomarkers (per modality)
        selection_results: dict[str, Any] = {}
        for modality in modalities:
            sel_result = await self._call_tool(
                "select_biomarkers",
                target=target,
                modality=modality,
                integration=config.get("integration", "union_weighted"),
                n_top=config.get("n_top", 30),
            )
            selection_results[modality] = sel_result
        report["steps"]["select_biomarkers"] = selection_results

        # Step 5: Run classification
        classify_result = await self._call_tool(
            "run_classification",
            target=config.get("classification_target", "mismatch"),
            phenotype_strategy=config.get("phenotype_strategy", "both"),
            cv_folds=config.get("cv_folds", 10),
        )
        report["steps"]["run_classification"] = classify_result

        # Step 6: Cross-omics matching
        match_result = await self._call_tool(
            "match_cross_omics_samples",
            dataset=config.get("dataset", "train"),
            distance_method=config.get("distance_method", "both"),
            n_iterations=config.get("n_iterations", 100),
            gene_sampling_fraction=config.get("gene_sampling_fraction", 0.8),
        )
        report["steps"]["match_cross_omics"] = match_result

        # Step 7: Explain top features
        all_genes: list[str] = []
        for modality_result in selection_results.values():
            for biomarker in modality_result.get("biomarkers", []):
                gene = biomarker.get("gene")
                if gene and gene not in all_genes:
                    all_genes.append(gene)

        if all_genes:
            explain_result = await self._call_tool(
                "explain_features",
                genes=all_genes[:30],
                context=f"{target}_classification",
                include_provenance=True,
            )
            report["steps"]["explain_features"] = explain_result

        # Summary
        report["summary"] = {
            "n_samples": load_result.get("samples", 0),
            "n_features_selected": len(all_genes),
            "ensemble_f1": classify_result.get("ensemble_f1", 0.0),
            "n_mismatches": len(match_result.get("identified_mismatches", [])),
        }

        return report

File: web/src/lib/agent-skills/biomarker-discovery.ts
/**
 * Agent skill: End-to-end biomarker discovery workflow.
 * Migrated from agent_skills/biomarker_discovery.py.
 *
 * Orchestrates: load -> impute -> availability -> select -> classify -> match -> explain
 */

import * as mlClient from "@/lib/ml-client";
import Anthropic from "@anthropic-ai/sdk";

export interface BiomarkerDiscoveryConfig {
  dataset?: string;
  target?: string;
  modalities?: string[];
  impute_strategy?: string;
  availability_threshold?: number;
  integration?: string;
  n_top?: number;
  cv_folds?: number;
  distance_method?: string;
  n_iterations?: number;
  gene_sampling_fraction?: number;
  classification_target?: string;
  phenotype_strategy?: string;
  data_dir?: string;
  enable_slm_routing?: boolean;
}

export async function runBiomarkerDiscovery(
  config: BiomarkerDiscoveryConfig = {},
): Promise<Record<string, unknown>> {
  const target = config.target ?? "msi";
  const modalities = config.modalities ?? ["proteomics", "rnaseq"];
  const dataset = config.dataset ?? "train";
  const report: Record<string, unknown> = {
    target,
    modalities,
    steps: {},
  };
  const steps: Record<string, unknown> = {};

  // Step 1: Impute per modality
  const imputeResults: Record<string, unknown> = {};
  for (const modality of modalities) {
    imputeResults[modality] = await mlClient.impute({
      dataset,
      modality,
      strategy: config.impute_strategy ?? "nmf",
      classify_missingness: true,
    });
  }
  steps.impute_missing = imputeResults;

  // Step 2: Select biomarkers per modality
  const selectionResults: Record<string, unknown> = {};
  const allGenes: string[] = [];
  for (const modality of modalities) {
    const result = await mlClient.selectFeatures({
      target,
      modality,
      integration: config.integration ?? "union_weighted",
      n_top: config.n_top ?? 30,
    });
    selectionResults[modality] = result;
    for (const biomarker of result.biomarkers ?? []) {
      const gene = (biomarker as Record<string, unknown>).gene as string;
      if (gene && !allGenes.includes(gene)) allGenes.push(gene);
    }
  }
  steps.select_biomarkers = selectionResults;

  // Step 3: Classification
  const classifyResult = await mlClient.classify({
    target: config.classification_target ?? "mismatch",
    classifiers: "ensemble",
    phenotype_strategy: config.phenotype_strategy ?? "both",
    cv_folds: config.cv_folds ?? 10,
  });
  steps.run_classification = classifyResult;

  // Step 4: Cross-omics matching
  const matchResult = await mlClient.matchCrossOmics({
    dataset,
    distance_method: config.distance_method ?? "both",
    n_iterations: config.n_iterations ?? 100,
    gene_sampling_fraction: config.gene_sampling_fraction ?? 0.8,
  });
  steps.match_cross_omics = matchResult;

  // Step 5: Explain top features via Claude
  if (allGenes.length > 0) {
    const client = new Anthropic();
    const message = await client.messages.create({
      model: "claude-sonnet-4-6",
      max_tokens: 2048,
      messages: [
        {
          role: "user",
          content: `Explain the biological relevance of these biomarker genes for ${target} classification in colorectal cancer: ${allGenes.slice(0, 30).join(", ")}. For each gene provide: function, ${target} relevance, pathways, clinical significance.`,
        },
      ],
    });
    steps.explain_features = {
      genes: allGenes.slice(0, 30),
      explanation:
        message.content[0].type === "text" ? message.content[0].text : "",
    };
  }

  report.steps = steps;
  report.summary = {
    n_features_selected: allGenes.length,
    ensemble_f1: classifyResult.ensemble_f1,
    n_mismatches: matchResult.identified_mismatches.length,
  };

  return report;
}
